### PR TITLE
Bus-mapping for opcode `calldatacopy`

### PR DESCRIPTION
**Pull request by silathdiir**
Associates issue https://github.com/appliedzkp/zkevm-circuits/issues/309

Only for code review, will implement `calldatacopy` bus-mapping on this branch to test if it is right as https://github.com/appliedzkp/zkevm-circuits/issues/309#issuecomment-1046799163

### CODE REVIEW

**Comment by HAOYUatHZ** [comment]
current status looks good to me. need to fix errors/conflicts introduced by recent merging

File: bus-mapping/src/circuit_input_builder.rs [hunk 1]
```diff
@@ -593,6 +636,27 @@ impl Transaction {
...
+#[derive(Debug)]
+/// Transaction configuration
+pub struct TransactionConfig {
+    /// If root call
+    pub is_root_call: bool,
+    /// Initialized length of call data
+    pub call_data_length: usize,
+    /// Initialized offset of call data
+    pub call_data_offset: u64,
+}
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:648]: remove

File: zkevm-circuits/src/test_util.rs [hunk 2]
```diff
@@ -29,22 +29,38 @@ pub fn get_fixed_table(conf: FixedTableConfig) -> Vec<FixedTableTag> {
 
 pub struct BytecodeTestConfig {
     pub enable_evm_circuit_test: bool,
-    pub evm_circuit_lookup_tags: Vec<FixedTableTag>,
     pub enable_state_circuit_test: bool,
+    pub is_root_call: bool,
```
Comment (icemelon) [inline zkevm-circuits/src/test_util.rs:33]: remove

File: zkevm-circuits/src/test_util.rs [hunk 3]
```diff
@@ -29,22 +29,38 @@ pub fn get_fixed_table(conf: FixedTableConfig) -> Vec<FixedTableTag> {
...
+    pub call_data_length: usize,
+    pub call_data_offset: u64,
```
Comment (icemelon) [inline zkevm-circuits/src/test_util.rs:35]: remove

**Review by icemelon — COMMENTED** [review]
after some further thoughts, I think we should leave the handling of calldatacopy for internal calls after the call PR https://github.com/appliedzkp/zkevm-circuits/pull/278 (busmapping included as well) is merged. The things we are adding in addition here are all temporary solutions and will be soon removed after call is supported.

File: bus-mapping/src/evm/opcodes/calldatacopy.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+use std::collections::HashMap;
+
+// The max number of bytes that can be copied in a step limited by the number
+// of cells in a step
+const MAX_COPY_BYTES: usize = 71;
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct Calldatacopy;
+
+impl Opcode for Calldatacopy {
```
Comment (DreamWuGit) [inline bus-mapping/src/evm/opcodes/calldatacopy.rs:16]: does need to add `CopyToMemory` bus mapping implementation  as in circuit it is a step gadget ?

File: bus-mapping/src/evm/opcodes.rs [hunk 5]
```diff
@@ -50,18 +52,30 @@ use stop::Stop;
...
+/// [`Operation`](crate::operation::Operation) that should be generated for one
+/// or multiple [`ExecStep`](crate::circuit_input_builder::ExecStep) depending
+/// of the [`OpcodeId`] it contains.
 pub trait Opcode: Debug {
     /// Generate the associated [`MemoryOp`](crate::operation::MemoryOp)s,
     /// [`StackOp`](crate::operation::StackOp)s, and
     /// [`StorageOp`](crate::operation::StorageOp)s associated to the Opcode
     /// is implemented for.
     fn gen_associated_ops(
         state: &mut CircuitInputStateRef,
```
Comment (DreamWuGit) [inline bus-mapping/src/evm/opcodes.rs:64]: first parameter `state` contains `step` member, could we reuse it instead of importing new one?

File: bus-mapping/src/evm/opcodes/calldatacopy.rs (hunk 4 above)
Comment (silathdiir) [reply bus-mapping/src/evm/opcodes/calldatacopy.rs:16]: It seems that `CallDataCopy` and `CopyToMemory` steps are generated from one GethExecStep. Since CopyToMemory steps are virtual. I added function `gen_memory_copy_steps` to generate CopyToMemory steps based on the previous `CallDataCopy` step.

File: bus-mapping/src/circuit_input_builder.rs [hunk 6]
```diff
@@ -654,21 +727,30 @@ pub struct CircuitInputStateRef<'a> {
     pub tx: &'a mut Transaction,
     /// Transaction Context
     pub tx_ctx: &'a mut TransactionContext,
-    /// Step
-    pub step: &'a mut ExecStep,
 }
 
 impl<'a> CircuitInputStateRef<'a> {
+    ///
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:733]:
```suggestion
    /// Create a new step from a `GethExecStep`
```

File: bus-mapping/src/circuit_input_builder.rs [hunk 7]
```diff
@@ -742,6 +831,14 @@ impl<'a> CircuitInputStateRef<'a> {
             .map(|call_idx| &self.tx.calls[call_idx])
     }
 
+    // pub fn caller(&self) -> Option<&Call> {
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:834]: remove this part for now

File: bus-mapping/src/circuit_input_builder.rs [hunk 8]
```diff
@@ -101,11 +101,73 @@ pub enum ExecError {
...
+    /// Virtual step Begin Tx
+    BeginTx,
+    /// Virtual step End Tx
+    EndTx,
+    /// Virtual step Copy To Memory
+    CopyToMemory,
+}
+
+impl ExecState {
+    /// Returns `true` if the `OpcodeId` is a `PUSHn`.
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:118]:
```suggestion
    /// Returns `true` if the `ExecState` is an opcode and the opcode is a `PUSHn`.
```

File: bus-mapping/src/circuit_input_builder.rs [hunk 9]
```diff
@@ -101,11 +101,73 @@ pub enum ExecError {
...
+    pub fn is_push(&self) -> bool {
+        if let ExecState::Op(op) = self {
+            op.is_push()
+        } else {
+            false
+        }
+    }
+
+    /// Returns `true` if the `OpcodeId` is a `DUPn`.
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:127]: ditto

File: bus-mapping/src/circuit_input_builder.rs [hunk 10]
```diff
@@ -101,11 +101,73 @@ pub enum ExecError {
...
+    pub fn is_dup(&self) -> bool {
+        if let ExecState::Op(op) = self {
+            op.is_dup()
+        } else {
+            false
+        }
+    }
+
+    /// Returns `true` if the `OpcodeId` is a `SWAPn`.
```
Comment (icemelon) [inline bus-mapping/src/circuit_input_builder.rs:136]: ditto

File: bus-mapping/src/evm/opcodes.rs [hunk 11]
```diff
@@ -40,29 +42,41 @@ use stop::Stop;
...
-        next_steps: &[GethExecStep],
-    ) -> Result<(), Error>;
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error>;
+
+    // fn gen_associated_ops(
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes.rs:58]: remove the comments

File: bus-mapping/src/evm/opcodes/calldatacopy.rs [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    let mut selectors = vec![0u8; MAX_COPY_BYTES];
+    for (idx, selector) in selectors.iter_mut().enumerate() {
+        if idx < bytes_left {
+            *selector = 1;
+            let addr = src_addr + idx as u64;
+            let byte = if addr < src_addr_end {
+                if is_root {
+                    state.tx.input[addr as usize]
+                } else {
+                    // TODO, read caller memory
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes/calldatacopy.rs:105]:
```suggestion
                    // TODO: read caller memory
```

File: bus-mapping/src/evm/opcodes/calldatacopy.rs [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+                    // state.push_memory_op(
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes/calldatacopy.rs:106]: remove the comment part for now

File: bus-mapping/src/evm/opcodes/calldatacopy.rs [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+                    //     exec_step,
+                    //     RW::READ,
+                    //     (addr as usize).into(),
+                    //     bytes_map[&addr],
+                    // );
+                    0u8
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes/calldatacopy.rs:112]:
```suggestion
                    unreachable!()
```

File: bus-mapping/src/evm/opcodes/stackonlyop.rs [hunk 15]
```diff
@@ -15,35 +15,38 @@ pub(crate) struct StackOnlyOpcode<const N_POP: usize, const N_PUSH: usize>;
...
-        steps: &[GethExecStep],
-    ) -> Result<(), Error> {
-        let step = &steps[0];
-
-        // N_POP stack reads
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+        // N stack reads
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes/stackonlyop.rs:22]:
```suggestion

        // N_POP stack reads
```

File: zkevm-circuits/src/evm_circuit/execution/memory_copy.rs [hunk 16]
```diff
@@ -70,14 +69,14 @@ impl<F: Field> ExecutionGadget<F> for CopyToMemoryGadget<F> {
...
             // Read bytes[i] from Tx
-            cb.condition(from_tx.expr() * read_flag.clone(), |cb| {
-                cb.tx_context_lookup(
-                    tx_id.expr(),
-                    TxContextFieldTag::CallData,
-                    Some(src_addr.expr() + i.expr()),
-                    buffer_reader.byte(i),
-                )
-            });
+            // cb.condition(from_tx.expr() * read_flag.clone(), |cb| {
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/memory_copy.rs:72]: revert these comments

File: zkevm-circuits/src/evm_circuit/witness.rs [hunk 17]
```diff
@@ -283,6 +283,28 @@ pub enum StepAuxiliaryData {
     },
 }
 
+impl From<circuit_input_builder::StepAuxiliaryData> for StepAuxiliaryData {
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/witness.rs:286]: We can probably directly use `StepAuxiliaryData` from the `circuit_input_builder` to avoid such conversion

File: zkevm-circuits/src/test_util.rs [hunk 18]
```diff
@@ -28,35 +28,37 @@ pub fn get_fixed_table(conf: FixedTableConfig) -> Vec<FixedTableTag> {
     }
 }
 
+#[derive(Debug, Clone)]
 pub struct BytecodeTestConfig {
     pub enable_evm_circuit_test: bool,
-    pub evm_circuit_lookup_tags: Vec<FixedTableTag>,
     pub enable_state_circuit_test: bool,
     pub gas_limit: u64,
+    pub evm_circuit_lookup_tags: Vec<FixedTableTag>,
```
Comment (icemelon) [inline zkevm-circuits/src/test_util.rs:36]: revert this change

File: zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs [hunk 19]
```diff
@@ -234,148 +234,33 @@ mod test {
...
-                    code_source: CodeSource::Account(bytecode.hash),
-                    ..Default::default()
-                }],
-                steps,
-                ..Default::default()
-            }],
-            rws,
-            bytecodes: vec![bytecode],
+        let call_data = Some(rand_bytes(call_data_length));
+        println!("calldata = {:?}", call_data);
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:256]: remove print

File: zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs [hunk 20]
```diff
@@ -234,148 +234,33 @@ mod test {
...
-                    is_create: false,
-                    code_source: CodeSource::Account(bytecode.hash),
-                    ..Default::default()
-                }],
-                steps,
-                ..Default::default()
-            }],
-            rws,
-            bytecodes: vec![bytecode],
+        let call_data = Some(rand_bytes(call_data_length));
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:255]:
```suggestion
        let call_data = rand_bytes(call_data_length);
```

File: zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs [hunk 21]
```diff
@@ -234,148 +234,33 @@ mod test {
...
+        println!("calldata = {:?}", call_data);
+        let test_config = BytecodeTestConfig {
             ..Default::default()
         };
-        assert_eq!(run_test_circuit_incomplete_fixed_table(block), Ok(()));
+        assert_eq!(
+            test_circuits_using_bytecode(bytecode, test_config, call_data),
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:261]:
```suggestion
            test_circuits_using_bytecode(bytecode, test_config, Some(call_data)),
```

**Review by icemelon — APPROVED** [review]

**Comment by silathdiir** [comment]
moved to https://github.com/appliedzkp/zkevm-circuits/pull/393

**Comment by icemelon** [comment]
close and move to https://github.com/appliedzkp/zkevm-circuits/pull/393
